### PR TITLE
[WFLY-16793] Add support for identity propagation with OIDC

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.redhat.com/browse/WFLY-16793

Depends on https://github.com/wildfly/wildfly-core/pull/5389

### CODE REVIEW

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc [hunk 1]
```diff
@@ -209,3 +215,71 @@ your application's `web.xml` file, as shown in the example below:
...
+
+If your application secured with OpenID Connect invokes an EJB within the same deployment (e.g.,
+within the same WAR or EAR) or invokes an EJB in a separate deployment (e.g., across EARs)
+and you'd like to secure the EJB using a different security domain from your servlet,
+additional configuration will be needed to outflow the security identities established by
+the virtual security domain to another security domain.
+
+The `virtual-security-domain` resource allows you to specify that security identities
+established by a virtual security domain should automatically be outflowed to other
+security domains. A `virtual-security-domain` resource has two attributes, as
```
Comment (OndrejKotek) [inline docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc:237]: There is also `outflow-anonymous` related and available.

File: testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+
+/**
+ * Test class to hold the identity propagation scenarios involving a virtual security domain with OpenID Connect.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+@ServerSetup({ OidcIdentityPropagationTestCase.EJBDomainSetupOverride.class, OidcIdentityPropagationTestCase.PropagationSetup.class, OidcIdentityPropagationTestCase.KeycloakAndSubsystemSetup.class })
+public class OidcIdentityPropagationTestCase {
```
Comment (OndrejKotek) [inline testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java:112]: I suggest to add/update
- negative cases
  - where the outflow/trust is not configured on one of the sides
  - where the outflowed identity is not available in the trusting security domain
- some current positive cases with `outflow-anonymous` enabled + test(s) for outflow not possible resulting in anonymous identity used

File: docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc (hunk 1 above)
Comment (fjuma) [reply docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc]: Updated, thanks!

File: testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java (hunk 2 above)
Comment (fjuma) [reply testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java:115]: I've added additional tests cases for this.

**Comment by fjuma** [comment]
@OndrejKotek Thanks for the feedback! I've made the suggested changes.

File: testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+     *
+     * The servlet attempts to invoke the local EJB.
+     * The security domain that's being used to secure the EJB hasn't been configured to
+     * trust the virtual security domain.
+     */
+    @Test
+    @OperateOnDeployment(ANOTHER_SINGLE_DEPLOYMENT_LOCAL)
+    public void testServletToEjbSingleDeploymentOutflowNotPossibleTrustNotConfigured() throws Exception {
+        String expectedMessage = "anonymous,false"; // charlie won't be outflowed, the target security domain's current identity will be used instead
+        loginToApp(KeycloakConfiguration.CHARLIE, KeycloakConfiguration.CHARLIE_PASSWORD, HttpURLConnection.HTTP_OK, expectedMessage, true, url.toURI().resolve("anotherWhoAmI"));
```
Comment (OndrejKotek) [inline testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java:524]: IIUIC, another user should be used (e.g. Alice) as Charlie _doesn't exist in the security domain that's being used to secure the EJB_.

File: testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java (hunk 2 above)
Comment (fjuma) [reply testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java:115]: This was just meant to mimic the existing `outflow-anonymous` attribute for existing `security-domains`. If the target security domain already had an identity established, `outflow-anonymous` can result in the identity for that target security domain being cleared. It's described in the docs like this:

`Outflowing anonymous has the effect of clearing any identity already established for that domain.`

I found it a bit tricky to simulate this with the way the tests are set up but will try to take a look.

Comment (OndrejKotek) [reply testsuite/integration/elytron-oidc-client/src/test/java/org/wildfly/test/integration/elytron/oidc/client/propagation/OidcIdentityPropagationTestCase.java:115]: Ah, yes, I'm sorry I was too fast and forgot about that _clearing_ part, thank you.

**Comment by OndrejKotek** [comment]
I get some rare intermittent issues in QE CI, roughly once per 12 TS executions, RHEL 8 + AdoptioumJDK 11/17 / OpenJDK 17 so far:

```
java.lang.AssertionError: Unexpected result alice,false
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotConfigured(OidcIdentityPropagationTestCase.java:539)

java.lang.AssertionError: Unexpected result alice,false
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbInvocation(OidcIdentityPropagationTestCase.java:487)
	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentLocal(OidcIdentityPropagationTestCase.java:456)

java.lang.AssertionError: Unexpected result charlie,false
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotPossibleNonExistentIdentity(OidcIdentityPropagationTestCase.java:508)

java.lang.AssertionError: Unexpected result alice,false
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotConfigured(OidcIdentityPropagationTestCase.java:539)
```

No luck with finding a cause so far.

**Comment by fjuma** [comment]
> I get some rare intermittent issues in QE CI, roughly once per 12 TS executions, RHEL 8 + AdoptioumJDK 11/17 / OpenJDK 17 so far:
> 
> ```
> java.lang.AssertionError: Unexpected result alice,false
> 	at org.junit.Assert.fail(Assert.java:89)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
> 	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotConfigured(OidcIdentityPropagationTestCase.java:539)
> 
> java.lang.AssertionError: Unexpected result alice,false
> 	at org.junit.Assert.fail(Assert.java:89)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
> 	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbInvocation(OidcIdentityPropagationTestCase.java:487)
> 	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentLocal(OidcIdentityPropagationTestCase.java:456)
> 
> java.lang.AssertionError: Unexpected result charlie,false
> 	at org.junit.Assert.fail(Assert.java:89)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
> 	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotPossibleNonExistentIdentity(OidcIdentityPropagationTestCase.java:508)
> 
> java.lang.AssertionError: Unexpected result alice,false
> 	at org.junit.Assert.fail(Assert.java:89)
> 	at org.junit.Assert.assertTrue(Assert.java:42)
> 	at org.wildfly.test.integration.elytron.oidc.client.OidcBaseTest.loginToApp(OidcBaseTest.java:380)
> 	at org.wildfly.test.integration.elytron.oidc.client.propagation.OidcIdentityPropagationTestCase.testServletToEjbSingleDeploymentOutflowNotConfigured(OidcIdentityPropagationTestCase.java:539)
> ```
> 
> No luck with finding a cause so far.

@OndrejKotek Are there any more details in the log file that you could point me to?

**Comment by OndrejKotek** [comment]
> @OndrejKotek Are there any more details in the log file that you could point me to?

I'm trying to reproduce with trace logging enabled.

**Comment by fjuma** [comment]
Just FYI, I suspect the intermittent failures are likely due to an issue with test setup / cleanup but need to investigate further.

**Comment by fjuma** [comment]
Just rebased since some of the CI jobs were still failing with not being able to access classes introduced in https://github.com/wildfly/wildfly-core/pull/5389 which were part of the Core upgrade.

**Comment by fjuma** [comment]
@bstansberry @darranl Just FYI, CI has passed on this one so this is ready to be merged for Beta. Follow-up items are being tracked via [WFLY-17781](https://issues.redhat.com/browse/WFLY-17781),  [WFLY-17785](https://issues.redhat.com/browse/WFLY-17785), and [WFLY-17784](https://issues.redhat.com/browse/WFLY-17784).

**Comment by bstansberry** [comment]
Thanks @fjuma and @OndrejKotek

**Comment by sjoerd0301** [comment]
Will this also be available in wildfly 26.x.x, or only in the next wildfly 27.x.x?

**Comment by sjoerd0301** [comment]
So that means I'd need to upgrade to jakarta 10? Is there any way there could be a hotfix for 26.x?

**Comment by fjuma** [comment]
How are your EJBs packaged? For WildFly 26.x, the identity will be successfully propagated from the web layer to the EJB layer if the EJB is contained in the same WAR. However, if the EJB is located outside the WAR and packaged in an EAR then the identity won't be propagated.

**Comment by decarbonite** [comment]
I'm having issue with propagating the identity from Web tier to EJB, I keep getting "anonymous" when retrieving it through `SessionContext.getCallerPrincipal().getName();`
I'm on WildFly 28.0.1.Final

My EAR looks like this: 
- some.ear
   --lib
   --META-INF
   --some.jar <- contains EJBs
   --some.war <- doesn't contain EJBs just the webapp

I have some.war listed in elytron-oidc-client subsystem as a `<secure-deployment>`. As I upgrade to WF 28, I added `<virtual-security-domain>` to elytron subsystem and `trusted-virtual-security-domains` attribute to the security domains, it looks like this:

```
<subsystem xmlns="urn:wildfly:elytron:17.0" final-providers="combined-providers" disallowed-providers="OracleUcrypto">
            <providers>
                <aggregate-providers name="combined-providers">
                    <providers name="elytron"/>
                    <providers name="openssl"/>
                </aggregate-providers>
                <provider-loader name="elytron" module="org.wildfly.security.elytron"/>
                <provider-loader name="openssl" module="org.wildfly.openssl"/>
            </providers>
            <security-domains>
                <virtual-security-domain name="some.ear" outflow-security-domains="ApplicationDomain ManagementDomain">
                <security-domain name="ManagementDomain" default-realm="ManagementRealm" permission-mapper="default-permission-mapper" pre-realm-principal-transformer="lowercase-principal-transformer" trusted-virtual-security-domains="some.ear">
                    <realm name="ManagementRealm" role-decoder="groups-to-roles"/>
                    <realm name="local" role-mapper="super-user-mapper"/>
                </security-domain>
                <security-domain name="ApplicationDomain" default-realm="ApplicationRealm" permission-mapper="default-permission-mapper" pre-realm-principal-transformer="lowercase-principal-transformer" trusted-virtual-security-domains="some.ear">
                    <realm name="ApplicationRealm" role-decoder="groups-to-roles"/>
                    <realm name="local"/>
                </security-domain>
                ...
            </security-domains>
         ...
</subsystem>

```

ejb3 and undertow subsystems reference ApplicationDomain as usual:
```
<application-security-domains>
        <application-security-domain name="other" security-domain="ApplicationDomain"/>
</application-security-domains>
```
The issue is not consistent either, if I restart the server sometimes it's gone, restart again it is back. As if some race condition at the server startup happens that determines whether it will propagate user correctly or not.

Previously I was on WF 26 with Keycloud adapters which worked fine to secure the application but this started happening as I moved to the elytron-oidc-client subsystem instead.

Any thoughts?

**Pull request merged**